### PR TITLE
Replace channels&locks with actors & log with slog

### LINKED ISSUE #28: Make logging per bot
**Is your feature request related to a problem? Please describe.**
With the multibot PR implemented it is not quite clear which specific bot logs where created by as soon as more than one is connected.

**Describe the solution you'd like**
The logging solution should be able to make a distinction between different bots. As an example it should be possible to log to dynamically create a log file whenever a new bot connects and write all related logs to that file.

**Describe alternatives you've considered**
I have not considered any alternatives though we might have to switch to slog to make these changes possible.

### LINKED ISSUE #31: Bot sometimes skips songs when adding a new song just as the currently playing one is ending
**Describe the bug**
When adding a song when the currently playing song is ending, the added song gets skipped immediately.

**To Reproduce**
1. Wait for song to end
2. Add new song

**Expected behavior**
Newly added song should go to the end of the playlist

**Additional context**
Logs from @Jokler:
```rust
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/playlist.rs:19 - Adding Some("Children Of Bodom - Are You Dead Yet?") to playlist
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/playlist.rs:26 - Popping Some(Some("Story of The Year- The Antidote")) from playlist
Feb 07 21:09:59 jokler-arch pokebot[3916198]: DEBUG tokio-runtime-worker-0[140625360508672] src/bot/music.rs:212 - Sending message to TeamSpeak: Playing 'Story of The Year- The Antidote'
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/bot/music.rs:224 - Setting TeamsSpeak description to Currently playing 'Story of The Year- The Antidote'
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/audio_player.rs:199 - Setting pipeline state to null
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/audio_player.rs:169 - Setting location URI: https://r1---sn-4g5ednse.googlevideo.com/videoplayback?expire=1581127269&>
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/audio_player.rs:207 - Setting pipeline state to playing
Feb 07 21:09:59 jokler-arch pokebot[3916198]: DEBUG unnamed[140624843994880] src/audio_player.rs:282 - Pipeline transitioned from Null to Ready, with Playing pending
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/playlist.rs:26 - Popping Some(Some("Breaking Benjamin - Follow Me")) from playlist
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/bot/music.rs:316 - Advancing playlist
Feb 07 21:09:59 jokler-arch pokebot[3916198]: DEBUG tokio-runtime-worker-0[140625360508672] src/bot/music.rs:212 - Sending message to TeamSpeak: Playing 'Breaking Benjamin - Follow Me'
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/bot/music.rs:224 - Setting TeamsSpeak description to Currently playing 'Breaking Benjamin - Follow Me'
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO tokio-runtime-worker-0[140625360508672] src/audio_player.rs:199 - Setting pipeline state to null
Feb 07 21:09:59 jokler-arch pokebot[3916198]: INFO unnamed[140624843994880] src/audio_player.rs:245 - Sending state Stopped to application
```

### LINKED ISSUE #53: Bot can still get stuck during shutdown
**Describe the bug**
For unknown reasons the bot does not always shut down after a panic.

**To Reproduce**
1. Panic in a tokio worker thread.
2. ??
3. Bot does not shut down

**Expected behavior**
Bot should shut down cleanly no matter what the error was

**Additional context**
```
Jul 06 17:37:48 jokler-arch pokebot[1987662]: thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Canceled', /home/jokler/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.22/src/runtime/threadpool/mod.rs:269:9
Jul 06 17:37:48 jokler-arch pokebot[1987662]: thread 'tokio-runtime-worker-0' panicked at 'called `Result::unwrap()` on an `Err` value: SpawnError { is_shutdown: true }', /home/jokler/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.9/src/global.rs:172:5
```

### LINKED ISSUE #58: Reconnect automatically to server
**Is your feature request related to a problem? Please describe.**
Currently if the server goes down the bot needs to be manually restarted to reconnect to the server.

**Describe the solution you'd like**
The bot should automatically reconnect the server when it the server comes back online without any manual intervention.

### LINKED ISSUE #69: Investigate structuring the bot(s) using actors
**Is your feature request related to a problem? Please describe.**
Currently the TeamSpeak/GStreamer parts of the bot is a bit of a mess with Mutexes/RwLock mixed with async functions

**Describe the solution you'd like**
See if we can move the concept of the main bot and the music bots into their own actors to decouple them more.

**Describe alternatives you've considered**
Try to clean up the current code?

### PR DESCRIPTION
**Pull request by Jokler**
**Summary**
This small PR fixes/implements the following **bugs/features**

* Use actors instead of locks (#69)
* Use slog for logging (#28)
* Fixed: Reconnect automatically (#58)
* Probably fixed: bot sometimes skips songs (#31)
* Maybe fixed: bot can still get stuck on shutdown (#53)

In the future we could investigate if `slog_scope` is viable for more of the logging we do.

**Closing issues**
Resolves #69.
Resolves #28.
Resolves #58.
Resolves #31.
Resolves #53.

### CODE REVIEW

File: src/audio_player.rs [hunk 1]
```diff
@@ -111,26 +95,31 @@ impl AudioPlayer {
...
+            volume_f64: 0.0,
             volume,
-            sender,
-            currently_playing: RwLock::new(None),
+            currently_playing: None,
         })
     }
 
-    fn create_audio_bin(
+    pub fn setup_with_callback(
```
Comment (fkaa) [inline src/audio_player.rs]: Setup what with callback?

File: src/audio_player.rs [hunk 2]
```diff
@@ -160,49 +149,64 @@ impl AudioPlayer {
...
-        let mut currently_playing = self.currently_playing.write().unwrap();
-        *currently_playing = Some(data);
+        self.currently_playing = Some(data);
 
         Ok(())
     }
 
     fn set_source_url(&self, location: String) -> Result<(), AudioPlayerError> {
-        info!("Setting location URI: {}", location);
+        info!(self.logger, "Setting source"; "uri" => &location);
```
Comment (fkaa) [inline src/audio_player.rs]: uri or url? :-)

File: src/bot/master.rs [hunk 3]
```diff
@@ -47,168 +57,121 @@ impl MasterBot {
...
+        trace!(logger, "Spawned master bot actor");
 
-        (bot, msg_loop)
+        bot_addr
     }
 
-    async fn build_bot_args_for(&self, id: ClientId) -> Result<MusicBotArgs, BotCreationError> {
-        let mut cteamspeak = self.teamspeak.clone();
-        let channel = match cteamspeak.channel_of_user(id).await {
+    async fn build_bot_args_for(
```
Comment (fkaa) [inline src/bot/master.rs]: for who?

Comment (fkaa) [reply src/bot/master.rs]: s/build/get/?

File: src/bot/master.rs [hunk 4]
```diff
@@ -47,168 +57,121 @@ impl MasterBot {
...
-            music_bots.connected_bots.remove(&n);
-            music_bots.available_names.push(name_index);
-            music_bots.available_ids.push(id_index);
-        });
-
-        info!("Connecting to {} on {}", channel_path, self.config.address);
+        info!(
+            self.logger,
+            "Connecting"; "channel" => &channel_path, "address" => &self.config.address
+        );
```
Comment (fkaa) [inline src/bot/master.rs]: Shouldn't this be before connecting, not returning the config?

File: src/bot/master.rs [hunk 5]
```diff
@@ -47,168 +57,121 @@ impl MasterBot {
...
+            local: false,
             channel: channel_path,
             verbose: self.config.verbose,
-            disconnect_cb,
+            logger: self.logger.new(o!("musicbot" => name)),
         })
     }
 
-    async fn spawn_bot_for(&self, id: ClientId) {
+    async fn spawn_bot_for(&mut self, id: ClientId) {
```
Comment (fkaa) [inline src/bot/master.rs]: for who?

File: src/bot/master.rs [hunk 6]
```diff
@@ -219,66 +182,114 @@ impl MasterBot {
         Ok(())
     }
 
-    async fn my_id(&self) -> ClientId {
-        let mut cteamspeak = self.teamspeak.clone();
+    pub async fn bot_data(&self, name: String) -> Option<crate::web_server::BotData> {
```
Comment (fkaa) [inline src/bot/master.rs:181]: get_bot_data

File: src/bot/master.rs [hunk 7]
```diff
@@ -219,66 +182,114 @@ impl MasterBot {
...
-            currently_playing: bot.currently_playing(),
-            playlist: bot.playlist_to_vec(),
-        })
-    }
-
-    pub fn bot_datas(&self) -> Vec<crate::web_server::BotData> {
-        let music_bots = self.music_bots.read().unwrap();
-
-        let len = music_bots.connected_bots.len();
+    pub async fn bot_datas(&self) -> Vec<crate::web_server::BotData> {
```
Comment (fkaa) [inline src/bot/master.rs:187]: get

File: src/bot/master.rs [hunk 8]
```diff
@@ -219,66 +182,114 @@ impl MasterBot {
...
-
+        for bot in self.connected_bots.values() {
+            let bot_data = bot.send(GetBotData).await.unwrap();
             result.push(bot_data);
         }
 
         result
     }
 
     pub fn bot_names(&self) -> Vec<String> {
```
Comment (fkaa) [inline src/bot/master.rs:198]: get

File: src/bot/master.rs [hunk 9]
```diff
@@ -219,66 +182,114 @@ impl MasterBot {
...
+        self.available_names.push(name);
+        self.available_ids.push(id);
+    }
+
+    pub async fn quit(&mut self, reason: String) -> Result<(), tsclientlib::Error> {
+        let futures = self
+            .connected_bots
+            .values()
+            .map(|b| b.send(Quit(reason.clone())));
+        for res in futures::future::join_all(futures).await {
```
Comment (fkaa) [inline src/bot/master.rs]: use `futures::future`

File: src/bot/music.rs [hunk 10]
```diff
@@ -59,112 +60,87 @@ pub enum MusicBotMessage {
...
-                    let mut rt = tokio::runtime::Runtime::new().unwrap();
-                    rt.block_on(cconnection.send_audio_packet(samples));
-                })),
-            )
-            .unwrap();
-
-            (audio_player, Some(connection))
-        };
-
+    pub async fn new(args: MusicBotArgs) -> Address<Self> {
```
Comment (fkaa) [inline src/bot/music.rs]: do we change to `MusicBot::spawn(args)` maybe? since we return the spawned actor's address now

File: src/bot/music.rs [hunk 11]
```diff
@@ -59,112 +60,87 @@ pub enum MusicBotMessage {
...
 
-        spawn_gstreamer_thread(player.clone(), tx.clone());
+        let playlist = Playlist::new(args.logger.clone());
 
-        if args.local {
-            spawn_stdin_reader(tx);
-        }
+        let teamspeak = if args.local {
+            info!(args.logger, "Starting in CLI mode");
+            debug!(args.logger, "Spawning stdin reader thread");
```
Comment (fkaa) [inline src/bot/music.rs]: no we're not :-o

File: src/bot/music.rs [hunk 12]
```diff
@@ -184,25 +160,16 @@ impl MusicBot {
         self.player.play().unwrap();
     }
 
-    pub async fn add_audio(&self, url: String, user: String) {
-        match crate::youtube_dl::get_audio_download_from_url(url).await {
+    pub async fn add_audio(&mut self, url: String, user: String) {
+        match crate::youtube_dl::get_audio_download_from_url(url, &self.logger).await {
```
Comment (fkaa) [inline src/bot/music.rs]: use namespace

File: src/bot/music.rs [hunk 13]
```diff
@@ -357,54 +302,47 @@ impl MusicBot {
...
             State::EndOfStream => format!("🎵 {} ({}%)", self.name, volume),
             _ => format!("🎵 {} - {} ({}%)", self.name, state, volume),
         };
         self.set_nickname(name).await;
     }
 
-    async fn on_state(&self, state: State) -> Result<(), AudioPlayerError> {
-        let current_state = *self.state.read().unwrap();
-        if current_state != state {
+    async fn on_state(&mut self, state: State) -> Result<(), AudioPlayerError> {
```
Comment (fkaa) [inline src/bot/music.rs]: s/state/new_state/g to make code clearer

File: src/bot/music.rs [hunk 14]
```diff
@@ -458,60 +393,163 @@ impl MusicBot {
         Ok(())
     }
 
-    async fn on_client_left_channel(&self, old_channel: ChannelId) {
-        let my_channel = self.my_channel().await;
+    // TODO logs an error if this music bot is the one leaving
```
Comment (fkaa) [inline src/bot/music.rs]: TODO now or later?

File: src/main.rs [hunk 15]
```diff
@@ -127,53 +166,97 @@ async fn run() -> Result<(), Box<dyn std::error::Error>> {
...
+            identity,
             channel: String::from("local"),
             verbose: bot_args.verbose,
-            disconnect_cb,
+            logger: root_logger,
         };
-        MusicBot::new(bot_args).await.1.await;
+        MusicBot::new(bot_args).await;
+
+        ctrl_c.await??;
```
Comment (fkaa) [inline src/main.rs:185]: ??!

File: src/main.rs [hunk 16]
```diff
@@ -127,53 +166,97 @@ async fn run() -> Result<(), Box<dyn std::error::Error>> {
...
         let bind_address = bot_args.bind_address.clone();
-        let (bot, fut) = MasterBot::new(bot_args).await;
+        let bot_name = bot_args.master_name.clone();
+        let bot_logger = root_logger.new(o!("master" => bot_name.clone()));
+        let bot = MasterBot::new(bot_args, bot_logger).await;
+
+        let weak_bot = bot.downgrade();
 
-        thread::spawn(|| {
+        let clogger = root_logger.clone();
```
Comment (fkaa) [inline src/main.rs]: lol

File: src/main.rs [hunk 17]
```diff
@@ -127,53 +166,97 @@ async fn run() -> Result<(), Box<dyn std::error::Error>> {
...
+        thread::spawn(move || {
```
Comment (fkaa) [inline src/main.rs]: should probably move into a function `spawn_web_server`

File: src/main.rs [hunk 18]
```diff
@@ -127,53 +166,97 @@ async fn run() -> Result<(), Box<dyn std::error::Error>> {
...
+            }
+        };
+
+        #[cfg(windows)]
+        ctrl_c.await??;
+
+        bot.send(Quit(String::from("Stopping")))
+            .await
+            .unwrap()
+            .unwrap();
```
Comment (fkaa) [inline src/main.rs]: We should look into anyhow or something so we can ?? from everywhere :-)

File: src/main.rs [hunk 19]
```diff
@@ -127,53 +166,97 @@ async fn run() -> Result<(), Box<dyn std::error::Error>> {
...
+
+pub async fn hangup() -> std::io::Result<()> {
+    signal(SignalKind::hangup())?.recv().await;
+    Ok(())
+}
+
+pub async fn quit() -> std::io::Result<()> {
+    signal(SignalKind::quit())?.recv().await;
+    Ok(())
+}
```
Comment (fkaa) [inline src/main.rs]: Shouldn't these be `#[cfg(unix)]` as well?

File: src/teamspeak/mod.rs [hunk 20]
```diff
@@ -86,50 +85,82 @@ fn get_message(event: &Event) -> Option<MusicBotMessage> {
...
+                .expect("connect_for_bot was called")
+                .with_connection(|mut conn| {
+                    conn.get_state()
+                        .expect("can get state")
+                        .server
+                        .set_subscribed(true)
+                        .send(&mut conn)
+                })
+                .await
+                .and_then(|v| v)
```
Comment (fkaa) [inline src/teamspeak/mod.rs:159]: isn't this a no-op?

File: src/teamspeak/mod.rs [hunk 21]
```diff
@@ -199,11 +231,15 @@ impl TeamSpeakConnection {
...
             })
             .await
-            .unwrap()
+            .map_err(|e| error!(self.logger, "Failed to get channel path of user"; "error" => %e))
+            .ok()
+            .and_then(|v| v)
     }
 
-    pub async fn my_channel(&mut self) -> ChannelId {
+    pub async fn my_channel(&mut self) -> Option<ChannelId> {
```
Comment (fkaa) [inline src/teamspeak/mod.rs]: current_channel?

File: src/teamspeak/mod.rs [hunk 22]
```diff
@@ -241,88 +282,90 @@ impl TeamSpeakConnection {
...
                     .get(&state.own_client)
                     .expect("can get myself")
                     .edit()
                     .set_description(&desc)
                     .send(&mut conn)
-                    .map_err(|e| error!("Failed to change description: {}", e));
             })
             .await
-            .unwrap()
+            .and_then(|v| v)
```
Comment (fkaa) [inline src/teamspeak/mod.rs:319]: big if, not sure if we can make these better, maybe get better at propagating errors?

File: src/teamspeak/mod.rs [hunk 23]
```diff
@@ -241,88 +282,90 @@ impl TeamSpeakConnection {
...
-
-    pub async fn disconnect(&mut self, reason: &str) {
+    pub async fn disconnect(&mut self, reason: &str) -> Result<(), tsclientlib::Error> {
         let opt = DisconnectOptions::new()
             .reason(Reason::Clientdisconnect)
             .message(reason);
-        self.handle.disconnect(opt).await.unwrap();
+        self.handle
+            .as_mut()
+            .expect("connect_for_bot was called")
```
Comment (fkaa) [inline src/teamspeak/mod.rs:367]: wrong expect message?

File: src/youtube_dl.rs [hunk 24]
```diff
@@ -28,13 +28,16 @@ where
     Ok(dur.map(Duration::from_secs_f64))
 }
 
-pub async fn get_audio_download_from_url(uri: String) -> Result<AudioMetadata, String> {
+pub async fn get_audio_download_from_url(
+    uri: String,
+    logger: &Logger,
+) -> Result<AudioMetadata, String> {
     //youtube-dl sometimes just fails, so we give it a second try
```
Comment (fkaa) [inline src/youtube_dl.rs:35]: fuckywucky 😠

File: src/youtube_dl.rs [hunk 25]
```diff
@@ -28,13 +28,16 @@ where
     Ok(dur.map(Duration::from_secs_f64))
 }
 
-pub async fn get_audio_download_from_url(uri: String) -> Result<AudioMetadata, String> {
+pub async fn get_audio_download_from_url(
```
Comment (fkaa) [inline src/youtube_dl.rs:31]: get_audio_metadata

Comment (fkaa) [reply src/youtube_dl.rs:31]: I'll just make a PR after this one 🦺

**Review by fkaa — CHANGES_REQUESTED** [review]
Skipped web server stuff cause I don't like it.

File: src/bot/master.rs (hunk 3 above)
Comment (Jokler) [reply src/bot/master.rs]: for the ClientId

File: src/bot/master.rs (hunk 5 above)
Comment (Jokler) [reply src/bot/master.rs]: again for the ClientId

File: src/teamspeak/mod.rs (hunk 20 above)
Comment (Jokler) [reply src/teamspeak/mod.rs:159]: it flattens `Result<Result<T, E>, E>`

Comment (Jokler) [reply src/teamspeak/mod.rs:159]: std will do the same: https://doc.rust-lang.org/src/core/result.rs.html#1211

File: src/bot/master.rs (hunk 5 above)
Comment (fkaa) [reply src/bot/master.rs]: then do s/spawn_bot_for/spawn_bot_for_client/?

File: src/audio_player.rs (hunk 1 above)
Comment (Jokler) [reply src/audio_player.rs]: Well `Self` obviously, but I can change it to mention what kind of callback it is.

Comment (Jokler) [reply src/audio_player.rs:106]: I guess it is not very clear that it has to be called either.

Comment (Jokler) [reply src/audio_player.rs]: I named it `setup_with_audio_callback` now.

File: src/audio_player.rs (hunk 2 above)
Comment (Jokler) [reply src/audio_player.rs]: I think it is URL for now, I believe URL is a subclass of URI which is limited to http but I might be wrong.

File: src/bot/master.rs (hunk 6 above)
Comment (Jokler) [reply src/bot/master.rs:181]: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

File: src/bot/music.rs (hunk 10 above)
Comment (Jokler) [reply src/bot/music.rs]: I think so.

File: src/bot/music.rs (hunk 11 above)
Comment (Jokler) [reply src/bot/music.rs]: If we log that later we have to clone the logger (KV map & sender). :/

File: src/bot/music.rs (hunk 14 above)
Comment (Jokler) [reply src/bot/music.rs]: Dunno how to solve it yet. So maybe later? ¯\\\_(ツ)\_/¯

File: src/main.rs (hunk 18 above)
Comment (Jokler) [reply src/main.rs]: Another PR.

File: src/teamspeak/mod.rs (hunk 22 above)
Comment (Jokler) [reply src/teamspeak/mod.rs:319]: Yes, in another PR.

File: src/teamspeak/mod.rs (hunk 23 above)
Comment (Jokler) [reply src/teamspeak/mod.rs:367]: nope

File: src/main.rs (hunk 16 above)
Comment (Jokler) [reply src/main.rs]: It is gone after creating the `spawn_web_server` function.

File: src/bot/music.rs (hunk 11 above)
Comment (Jokler) [reply src/bot/music.rs]: Moved it anyway.

**Comment by Jokler** [comment]
Nothing really changed in the webserver files anyway, it just uses the xtra actors instead of actix actors now.

File: src/bot/music.rs (hunk 14 above)
Comment (Jokler) [reply src/bot/music.rs]: Renamed it to FIXME.

File: src/log_bridge.rs [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+
+impl<T: log::Log> Drain for LogBridge<T> {
+    type Ok = ();
+    type Err = slog::Error;
+
+    fn log(&self, record: &slog::Record, kvs: &slog::OwnedKVList) -> Result<(), Self::Err> {
+        let mut target = record.tag();
+        if target.is_empty() {
+            target = record.module();
+        }
```
Comment (Jokler) [inline src/log_bridge.rs:16]: I am not sure if this is how `log` behaves or if it should be kept empty.

Comment (Jokler) [reply src/log_bridge.rs:16]: Looks like it is https://docs.rs/log/0.4.11/src/log/macros.rs.html#54

**Review by fkaa — APPROVED** [review]

**Pull request merged**